x-pack/plugins/fleet/server/services/metrics/fleet_metrics_task.ts

```diff
@@ -0,0 +1,182 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type {
+  ConcreteTaskInstance,
+  TaskManagerStartContract,
+  TaskManagerSetupContract,
+} from '@kbn/task-manager-plugin/server';
+import { throwUnrecoverableError } from '@kbn/task-manager-plugin/server';
+import type { ElasticsearchClient } from '@kbn/core/server';
+import { withSpan } from '@kbn/apm-utils';
+
+import { appContextService } from '../app_context';
+
+import type { AgentMetrics } from './fetch_agent_metrics';
+
+export class FleetMetricsTask {
+  private taskManager?: TaskManagerStartContract;
+  private taskVersion = '0.0.8';
+  private taskType = 'Fleet-Metrics-Task';
+  private wasStarted: boolean = false;
+  private interval = '1m';
+  private timeout = '1m';
+  private abortController = new AbortController();
+  private esClient?: ElasticsearchClient;
+
+  constructor(
+    taskManager: TaskManagerSetupContract,
+    fetchAgentMetrics: (abortController: AbortController) => Promise<AgentMetrics | undefined>
+  ) {
+    taskManager.registerTaskDefinitions({
+      [this.taskType]: {
+        title: 'Fleet Metrics Task',
+        timeout: this.timeout,
+        maxAttempts: 1,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return withSpan({ name: this.taskType, type: 'metrics' }, () =>
+                this.runTask(taskInstance, () => fetchAgentMetrics(this.abortController))
+              );
+            },
+
+            cancel: async () => {
+              this.abortController.abort('task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  private runTask = async (
+    taskInstance: ConcreteTaskInstance,
+    fetchAgentMetrics: () => Promise<AgentMetrics | undefined>
+  ) => {
+    if (!this.wasStarted) {
+      appContextService.getLogger().debug('[runTask()] Aborted. Task not started yet');
+      return;
+    }
+    // Check that this task is current
+    if (taskInstance.id !== this.taskId) {
+      throwUnrecoverableError(new Error('Outdated task version for task: ' + taskInstance.id));
+      return;
+    }
+    if (!this.esClient) {
+      appContextService.getLogger().info('esClient not set, skipping Fleet metrics task');
+      return;
+    }
+    appContextService.getLogger().info('Running Fleet metrics task');
+
+    try {
+      const agentMetrics = await fetchAgentMetrics();
+      if (!agentMetrics) {
+        return;
+      }
+      const { agents_per_version: agentsPerVersion, agents } = agentMetrics;
+      const clusterInfo = await this.esClient.info();
+      const getCommonFields = (dataset: string) => {
+        return {
+          data_stream: {
+            dataset,
+            type: 'metrics',
+            namespace: 'default',
```

Comment:
should we use the current space here?

Reply:
Does the current space affect the agent documents' visibility? I think all the data here is space-agnostic, so namespacing by Kibana space seems incorrect. Maybe I am misunderstanding though.

Reply:
I'll leave as default for now, thanks.